### PR TITLE
[FIX] website: not move images twice within image wall

### PR DESCRIPTION
**Pull request by bso-odoo**
Since [1] the "Images Add/Remove All" buttons are on top of the background options in order to appear as the first option for the "Image Wall" snippet.
This makes the JS related to the handling of the options of that snippet created twice: once for the buttons above and once for the options below.
Because of this, events are registered by both instances and they both get notified on option update. Therefore, when an image is moved, it is moved twice instead of once.

This commit differentiates both instances, and makes the one that handles the "Add" and "Remove All" buttons be the only one that works on the images.
It also replaces the way [2] solved a double opening of the image selection dialog.
In stable the differentiation is done with conditional statements.
In master the instances should be of distinct classes.

It also neutralizes the decimal images sizes which make the visual
sequence of images confusing.

[1]: https://github.com/odoo/odoo/commit/b6494fcf284edcddaa36b5fcfd407a6d7186fbd7
[2]: https://github.com/odoo/odoo/commit/5738062be1a511cf09512af7d330da5f7d22aac0

task-2990053

### CODE REVIEW

**Comment by qsm-odoo** [comment]
@robodoo delegate=Guillaume-gdi

File: addons/website/static/src/snippets/s_image_gallery/options.js [hunk 1]
```diff
@@ -16,36 +16,41 @@ options.registry.gallery = options.Class.extend({
...
+                        });
                     });
-                });
+                }
+            });
+        } else {
+            const $container = this.$('> .container, > .container-fluid, > .o_container_small');
+            if ($container.find('> *:not(div)').length) {
+                self.mode(null, self.getMode());
             }
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_image_gallery/options.js:53]: I don't understand why do you need to ignore the event listeners if `this.hasAddImages`. I took your solution but without your changes in start except `this.hasAddImages = this.$el[0].querySelector('we-button[data-add-images]');` and it solve the issue. Why the rest of start changes are needed ?

Also if it is needed, I would keep the diff minimal by doing something like
```js
if (!this.hasAddImages) {
    const $container = this.$('> .container, > .container-fluid, > .o_container_small');
    if ($container.find('> *:not(div)').length) {
        self.mode(null, self.getMode());
    }
    return this._super.apply(this, arguments); 
}
// Make sure image previews are updated if images are changed
this.$target.on('image_changed.gallery', 'img', function (ev) {
...
return this._super.apply(this, arguments); 
```

File: addons/website/static/src/snippets/s_image_gallery/options.js [hunk 2]
```diff
@@ -92,11 +97,6 @@ options.registry.gallery = options.Class.extend({
      * @see this.selectClass for parameters
      */
     addImages: function (previewMode) {
-        // Prevent opening dialog twice.
-        if (this.__imageDialogOpened) {
-            return Promise.resolve();
-        }
-        this.__imageDialogOpened = true;
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_image_gallery/options.js:99]: Maybe the revert could be a different commit ? :thinking:

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]

File: addons/website/static/src/snippets/s_image_gallery/options.js (hunk 1 above)
Comment (bso-odoo) [reply addons/website/static/src/snippets/s_image_gallery/options.js]: It is not strictly needed to make it work - but not doing it makes those events triggered several times.
Doing it also prepares the split into two distinct classes that will be done in master.

I followed your advice to make the diff smaller.

File: addons/website/static/src/snippets/s_image_gallery/options.js [hunk 3]
```diff
@@ -197,6 +203,8 @@ options.registry.gallery = options.Class.extend({
             _.each(cols, function (col) {
                 var $col = $(col);
                 var height = $col.is(':empty') ? 0 : $col.find('img').last().offset().top + $col.find('img').last().height() - self.$target.offset().top;
+                // Neutralize invisible sub-pixel height differences.
+                height = Math.round(height);
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_image_gallery/options.js:207]: Can you explain what does it fix ? Maybe add screenshots in PR description :eyes:

File: addons/website/static/src/snippets/s_image_gallery/options.js [hunk 4]
```diff
@@ -302,12 +310,20 @@ options.registry.gallery = options.Class.extend({
...
             data.$image.remove(); // Force the removal of the image before reset
             this.mode('reset', this.getMode());
         } else if (name === 'image_index_request') {
             var imgs = this._getImages();
             var position = _.indexOf(imgs, data.$image[0]);
+            if (position === 0 && data.position === 'prev') {
+                data.position = 'last';
+            } else if (position === imgs.length - 1 && data.position === 'next') {
+                data.position = 'first';
+            }
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_image_gallery/options.js:326]: I think I'm missing something :thinking: Why is it needed now ? How it was working before ?

File: addons/website/static/src/snippets/s_image_gallery/options.js [hunk 5]
```diff
@@ -16,6 +16,17 @@ options.registry.gallery = options.Class.extend({
      */
     start: function () {
         var self = this;
+        // TODO In master: define distinct classes.
+        // Differentiate both instances of this class.
+        this.hasAddImages = this.$el[0].querySelector('we-button[data-add-images]');
+
+        if (!this.hasAddImages) {
+            const $container = this.$('> .container, > .container-fluid, > .o_container_small');
+            if ($container.find('> *:not(div)').length) {
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_image_gallery/options.js:25]: You could take to opportunity to remove JQuery :+1:

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]
Also, can you add a test to make sure the bug doesn't come back? Thanks

File: addons/website/static/src/snippets/s_image_gallery/options.js (hunk 4 above)
Comment (bso-odoo) [reply addons/website/static/src/snippets/s_image_gallery/options.js:326]: Because it was broken.
E.g.:
- select first image and move it to the previous position. => It reached the position before the expected one.
- select last image and move it to the next position. => Nothing happened.

File: addons/website/static/src/snippets/s_image_gallery/options.js (hunk 3 above)
Comment (bso-odoo) [reply addons/website/static/src/snippets/s_image_gallery/options.js:207]: Screenshots will not show anything because it is a sub-pixel issue.
Using the default Image Wall images, the top one of the last column (writing on documents with a pen) is actually one pixel taller than the other ones. Once resized to the grid, they have the same height in number of pixels, but the column height with that image is 266.96... instead of 266.73... When putting the pictures in the column layout, each "next picture" is put inside the least tall column.
If you select the "writing" picture, and move it to the previous column, you will see that the 6th picture (the wine glass) moves together with it, because other columns are filled before that one by the previous pictures since they are less tall.
The fix makes the comparison happen on visible pixel differences.

Comment (Guillaume-gdi) [reply addons/website/static/src/snippets/s_image_gallery/options.js:207]: Ok I think this deserves it's own commit with your explanation and steps to reproduce :+1:

File: addons/website/static/src/snippets/s_image_gallery/options.js
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_image_gallery/options.js:1]: I noticed an issue when I
- Drop Images Wall
- Set the mode to float
- Select the writing picture
- Move it to previous position
- Set the mode to grid

Traceback :grimacing: 

Can you fix that to please

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]

File: addons/website/static/src/snippets/s_image_gallery/options.js [hunk 6]
```diff
@@ -197,6 +197,8 @@ options.registry.gallery = options.Class.extend({
             _.each(cols, function (col) {
                 var $col = $(col);
                 var height = $col.is(':empty') ? 0 : $col.find('img').last().offset().top + $col.find('img').last().height() - self.$target.offset().top;
+                // Neutralize invisible sub-pixel height differences.
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_image_gallery/options.js:206]: I'm not convinced by this commit, but in any case when I put the same images in 14.0, I have the same bug (the set to previous position moves the entire col). I suppose you can create an other PR that target 14.0, but if this is the expected behavior for images that have different height, isn't the fix about resize the default image ?

File: addons/website/static/tests/tours/snippet_images_wall.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+"use strict";
+
+const tour = require("web_tour.tour");
+const wTourUtils = require("website.tour_utils");
+
+tour.register("snippet_images_wall", {
+    test: true,
+    url: "/",
+}, [
+    ...wTourUtils.clickOnEditAndWaitEditMode(),
```
Comment (Guillaume-gdi) [inline addons/website/static/tests/tours/snippet_images_wall.js:11]: I would remove the `...wTourUtils.clickOnEditAndWaitEditMode()` and the `const tour = require("web_tour.tour");` and use the `wTourUtils.registerWebsitePreviewTour(` with `edition: true`

File: addons/website/static/tests/tours/snippet_images_wall.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    wTourUtils.dragNDrop({
+        id: "s_images_wall",
+        name: "Images Wall",
+}), wTourUtils.clickOnSnippet({
+    id: 's_image_gallery',
+    name: 'Images Wall',
+}), {
+    // Prefixing selectors with #wrap to avoid matching droppable block.
+    content: "Click on third image",
+    trigger: "#wrap .s_image_gallery img[data-index='2']",
```
Comment (Guillaume-gdi) [inline addons/website/static/tests/tours/snippet_images_wall.js:21]: I think the trigger should start with `iframe`

File: addons/website/static/tests/tours/snippet_images_wall.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    content: "Click on move to next",
+    extra_trigger: "#wrap .s_image_gallery .o_masonry_col:nth-child(3):has(img[data-index='5'][data-original-src*='sign'])",
+    trigger: ".snippet-option-gallery_img we-button[data-position='next']",
+}, {
+    content: "Click on move to last",
+    extra_trigger: "#wrap .s_image_gallery .o_masonry_col:nth-child(1):has(img[data-index='0'][data-original-src*='sign'])",
+    trigger: ".snippet-option-gallery_img we-button[data-position='last']",
+}, {
+    content: "Check layout",
+    trigger: "#wrap .s_image_gallery .o_masonry_col:nth-child(3):has(img[data-index='5'][data-original-src*='sign'])",
```
Comment (Guillaume-gdi) [inline addons/website/static/tests/tours/snippet_images_wall.js:43]: Same

File: addons/website/static/tests/tours/snippet_images_wall.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
+odoo.define("website.tour.snippet_images_wall", function (require) {
```
Comment (Guillaume-gdi) [inline addons/website/static/tests/tours/snippet_images_wall.js:1]: You should use the new way `/** @odoo-module **/`

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]

File: addons/website/static/tests/tours/snippet_images_wall.js (hunk 7 above)
Comment (Guillaume-gdi) [reply addons/website/static/tests/tours/snippet_images_wall.js:11]: Oups this does not exits yet in 15.0 MB.

File: addons/website/static/src/snippets/s_image_gallery/options.js (hunk 6 above)
Comment (bso-odoo) [reply addons/website/static/src/snippets/s_image_gallery/options.js:206]: Extracted to a 14.0 PR: https://github.com/odoo/odoo/pull/119064.
IMHO it has to be fixed there - the default image causing the error is just a random event that could have happened to any user after e.g. manual cropping.

File: addons/website/static/tests/tours/snippet_images_wall.js (hunk 10 above)
Comment (bso-odoo) [reply addons/website/static/tests/tours/snippet_images_wall.js:1]: Oops, I had noticed it, but then I somehow thought I was in 14.0. Doing it now.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 11]
```diff
@@ -5654,6 +5654,9 @@ registry.ImageTools = ImageHandlerOption.extend({
      * @override
      */
     _relocateWeightEl() {
+        if (!this.$overlay.data('$optionsSection')) {
+            return;
+        }
```
Comment (Guillaume-gdi) [inline addons/web_editor/static/src/js/editor/snippets.options.js:5659]: Why thy bug happens when the user goes from Float to Grid and not from Grid to float ? :thinking:

File: addons/website/static/tests/tours/snippet_images_wall.js [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    test: true,
+    url: "/",
+}, [
+    ...wTourUtils.clickOnEditAndWaitEditMode(),
+    wTourUtils.dragNDrop({
+        id: "s_images_wall",
+        name: "Images Wall",
+}), wTourUtils.clickOnSnippet({
+    id: 's_image_gallery',
+    name: 'Images Wall',
```
Comment (Guillaume-gdi) [inline addons/website/static/tests/tours/snippet_images_wall.js:16]: Quote consistency. Use double quotes everywhere :+1:

File: addons/website/tests/test_snippets.py [hunk 13]
```diff
@@ -53,3 +53,6 @@ def test_06_snippet_image_gallery(self):
             'url': base + '/web/image/website.s_banner_default_image.jpg',
         })
         self.start_tour("/", "snippet_image_gallery", login='admin')
+
+    def test_07_snippet_images_wall(self):
+        self.start_tour("/", "snippet_images_wall", login='admin')
```
Comment (Guillaume-gdi) [inline addons/website/tests/test_snippets.py:58]: Quote consistency, even if the rest of the file is mix of single and double, I think that in python we encourage single quote string

**Review by Guillaume-gdi — CHANGES_REQUESTED** [review]

File: addons/website/static/tests/tours/snippet_images_wall.js (hunk 7 above)
Comment (qsm-odoo) [reply addons/website/static/tests/tours/snippet_images_wall.js:10]: What about in 32MB?

Comment (Guillaume-gdi) [reply addons/website/static/tests/tours/snippet_images_wall.js:10]: I meant "my bad" :sweat_smile:

File: addons/website/static/src/snippets/s_image_gallery/options.js [hunk 14]
```diff
@@ -16,6 +16,17 @@ options.registry.gallery = options.Class.extend({
      */
     start: function () {
         var self = this;
+        // TODO In master: define distinct classes.
+        // Differentiate both instances of this class.
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_image_gallery/options.js:20]: I would be a bit more explicit here. Add that we don't want to register the event listeners twice :+1:

File: addons/website/static/src/snippets/s_image_gallery/options.js [hunk 15]
```diff
@@ -16,6 +16,17 @@ options.registry.gallery = options.Class.extend({
...
+        this.hasAddImages = this.el.querySelector('we-button[data-add-images]');
+
+        if (!this.hasAddImages) {
+            const containerEl = this.$target[0].querySelector(':scope > .container, :scope > .container-fluid, :scope > .o_container_small');
+            if (containerEl.querySelector(':scope > *:not(div)')) {
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_image_gallery/options.js:25]: Double quote

File: addons/website/static/src/snippets/s_image_gallery/options.js [hunk 16]
```diff
@@ -302,12 +308,20 @@ options.registry.gallery = options.Class.extend({
      */
     notify: function (name, data) {
         this._super(...arguments);
+        if (!this.hasAddImages) {
+            return;
+        }
```
Comment (Guillaume-gdi) [inline addons/website/static/src/snippets/s_image_gallery/options.js:313]: I would add comment here to explain that the widget is instantiate multiple times and we don't want to move the image multiple times, and also that it will be changed in master :+1:

**Review by Guillaume-gdi — APPROVED** [review]

**Comment by Guillaume-gdi** [comment]
@robodoo rebase-ff r+
